openssl/src/ssl/mod.rs

```diff
@@ -3445,6 +3445,39 @@ impl SslRef {
     pub fn security_level(&self) -> u32 {
         unsafe { ffi::SSL_get_security_level(self.as_ptr()) as u32 }
     }
+
+    /// Get the temporary key provided by the peer that is used during key
+    /// exchange.
+    // We use an owned value because EVP_KEY free need to be called when it is
+    // dropped
+    #[corresponds(SSL_get_peer_tmp_key)]
+    #[cfg(ossl300)]
+    pub fn peer_tmp_key(&self) -> Result<PKey<Public>, ErrorStack> {
+        unsafe {
+            let mut key = ptr::null_mut();
+            match cvt_long(ffi::SSL_get_peer_tmp_key(self.as_ptr(), &mut key)) {
+                Ok(_) => Ok(PKey::<Public>::from_ptr(key)),
+                Err(e) => Err(e),
+            }
+        }
+    }
+
+    /// Returns the temporary key from the local end of the connection that is
+    /// used during key exchange.
+    // We use an owned value because EVP_KEY free need to be called when it is
+    // dropped
+    #[corresponds(SSL_get_tmp_key)]
+    #[cfg(ossl300)]
+    pub fn tmp_key(&self) -> Result<PKey<Public>, ErrorStack> {
```

Comment:
Does the local side return private components or just public ones as well?

Reply:
Looks like you are correct, and there are private components:

https://github.com/openssl/openssl/blob/master/ssl/statem/extensions_clnt.c#L618
The s3.tmp.pkey pointer that this function returns is created from `ssl_generate_pkey_group` what has the following comment
> Generate a private key from a group ID

I'm not particularly familiar with this part of openssl, but seems safe to say that there are private components included. I'll go ahead and update the PR to switch the return type of `tmp_key` to `PKey<Private>`